python/mlc_chat/model/baichuan/baichuan_model.py

```diff
@@ -0,0 +1,244 @@
+                    "`context_window_size`, `max_position_embeddings` or `max_sequence_length` is "
+                    "provided in `config.json`."
+                )
+        if self.prefill_chunk_size == 0:
+            logger.info(
+                "%s defaults to %s (%d)",
+                bold("prefill_chunk_size"),
+                bold("context_window_size"),
+                self.context_window_size,
+            )
+            self.prefill_chunk_size = self.context_window_size
+        elif self.prefill_chunk_size > self.context_window_size:
+            logger.info(
+                "Overriding %s from %d to %d (%s)",
+                bold("prefill_chunk_size"),
+                self.prefill_chunk_size,
+                self.context_window_size,
+                bold("context_window_size"),
+            )
+            self.prefill_chunk_size = self.context_window_size
+        assert self.tensor_parallel_shards == 1, "Baichuan currently does not support sharding."
+
+
+class BaichuanAttention(nn.Module):  # pylint: disable=too-many-instance-attributes
+    def __init__(self, config: BaichuanConfig):
+        self.hidden_size = config.hidden_size
+        self.num_heads = config.num_attention_heads
+        self.head_dim = self.hidden_size // self.num_heads
+        self.max_position_embeddings = config.context_window_size
+
+        self.W_pack = nn.Linear(self.hidden_size, 3 * self.hidden_size, bias=False)
+        self.o_proj = nn.Linear(self.num_heads * self.head_dim, self.hidden_size, bias=False)
+
+        self.k_cache = nn.KVCache(config.context_window_size, [self.num_heads, self.head_dim])
+        self.v_cache = nn.KVCache(config.context_window_size, [self.num_heads, self.head_dim])
+
+    def forward(  # pylint: disable=too-many-locals
+        self,
+        hidden_states: Tensor,
+        attention_mask: Tensor,
+        total_seq_len: tir.Var,
+    ):
+        d, h, t = self.head_dim, self.num_heads, total_seq_len
+        b, s, _ = hidden_states.shape
+        assert b == 1, "Only support batch size 1 at this moment."
+        # Step 1. QKV Projection
+        qkv = self.W_pack(hidden_states)
+        qkv = op.reshape(qkv, (b, s, 3 * h, d))
+        # Step 2. Apply QK rotary embedding
+        q, k, v = op_ext.llama_rope(qkv, t, 10000, h, h)
+        # Step 3. Query and update KVCache
+        self.k_cache.append(op.squeeze(k, axis=0))
+        self.v_cache.append(op.squeeze(v, axis=0))
+        k = self.k_cache.view(t)
+        v = self.v_cache.view(t)
+        # Step 4. Compute softmax(Q @ K^T / sqrt(d)) @ V
+        output = op_ext.attention(q, k, v, casual_mask=attention_mask)
+        # Step 5. Apply output projection
+        return self.o_proj(output)
+
+
+class BaichuanMLP(nn.Module):
+    def __init__(self, config: BaichuanConfig):
+        self.gate_proj = nn.Linear(config.hidden_size, config.intermediate_size, bias=False)

```

Comment:
Here, we can combine the gate_proj and up_proj matmul similar to what we did in combing QKV in attention layer.
```
self.gate_up_proj = nn.Linear(
    in_features=config.hidden_size,
    out_features=2 * config.intermediate_size,
    bias=False,
)
self.down_proj = nn.Linear(self.intermediate_size, config.hidden_size, bias=False)
```

And in `forward`,
```
concat_x1_x2 = self.gate_up_proj(x)
x1, x2 = op.split(concat_x1_x2, 2, axis=-1)
return self.down_proj(self.act_fn(x1) * x2)
```

Also need to transform the parameters in the loader, similar to this one: https://github.com/mlc-ai/mlc-llm/blob/main/python/mlc_chat/model/llama/llama_loader.py#L65

Reply:
Got it. So it is usually better to fuse ops than use them independently?

Reply:
yes,

Reply:
I see. Thanks! I have already combined them in the code.